python/across_api/burstcube/api.py

```diff
@@ -1,3 +1,184 @@
 # Copyright © 2023 United States Government as represented by the
 # Administrator of the National Aeronautics and Space Administration.
 # All Rights Reserved.
+
+import gzip
+from typing import Annotated, BinaryIO, Tuple, Union
+
+from astropy.io import fits  # type: ignore
+from fastapi import Depends, File, HTTPException, Security, UploadFile, status
+
+from ..auth.api import claims, scope_authorize
+from ..base.depends import (
+    ErrorRadiusDep,
+    ExposureDep,
+    IdDep,
+    LimitDep,
+    OffsetDep,
+    OptionalDateRangeDep,
+    OptionalDurationDep,
+    OptionalRaDecDep,
+    TriggerTimeDep,
+)
+from ..base.api import app
+from .schema import BurstCubeTOORequestsSchema, BurstCubeTOOSchema, BurstCubeTriggerInfo
+from .toorequest import BurstCubeTOO, BurstCubeTOORequests
+
+
+def read_healpix_file(healpix_file: UploadFile) -> Tuple[fits.FITS_rec, str]:
+    """Read in a HEALPix file in FITS format and return the HDUList. Supports
+    gzipped FITS files"""
+
+    # Type hint for the file object
+    file: Union[gzip.GzipFile, BinaryIO]
+
+    # If the file is a gzip file, open it with gzip.open, otherwise just
+    # pass along the filehandle
+    if healpix_file.content_type == "application/x-gzip":
+        file = gzip.open(healpix_file.file, "rb")
+    else:
+        file = healpix_file.file
+
+    # Open HEALPix fits file and extract data and ordering scheme
+    try:
+        with fits.open(file) as hdu:
+            healpix_data = hdu[1].data
+            healpix_scheme = hdu[1].header["ORDERING"]
+    except (OSError, KeyError, IndexError):
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Invalid HEALPix file.",
+        )
+    return healpix_data, healpix_scheme
+
+
+@app.post(
+    "/burstcube/too",
+    status_code=status.HTTP_201_CREATED,
+    dependencies=[
+        Security(scope_authorize, scopes=["gcn.nasa.gov/kafka-public-consumer"])
+    ],
+)
+async def burstcube_too_submit(
+    credential: Annotated[dict, Depends(claims)],
+    ra_dec: OptionalRaDecDep,
+    error_radius: ErrorRadiusDep,
+    trigger_time: TriggerTimeDep,
+    trigger_info: BurstCubeTriggerInfo,
+    exposure: ExposureDep,
+    offset: OffsetDep,
+    healpix_file: UploadFile = File(
+        None, description="HEALPix file describing the localization."
+    ),
+) -> BurstCubeTOOSchema:
+    """
+    Resolve the name of an astronomical object to its coordinates.
+    """
+    # Construct the TOO object.
+    too = BurstCubeTOO(
+        username=credential["sub"],
```

Comment:
This field should be called `sub`, not `username`. OIDC subject claims are not usernames.

Reply:
Fixed